build_runner/lib/src/generate/build.dart

```diff
@@ -50,45 +54,58 @@ import 'watch_impl.dart' as watch_impl;
 /// the default [logLevel] to [Level.ALL] and removes stack frame folding, among
 /// other things.
 Future<BuildResult> build(List<BuilderApplication> builders,
-        {bool deleteFilesByDefault,
-        bool failOnSevere,
-        bool assumeTty,
-        String configKey,
-        PackageGraph packageGraph,
-        RunnerAssetReader reader,
-        RunnerAssetWriter writer,
-        Level logLevel,
-        onLog(LogRecord record),
-        Stream terminateEventStream,
-        bool enableLowResourcesMode,
-        Map<String, String> outputMap,
-        bool trackPerformance,
-        bool skipBuildScriptCheck,
-        bool verbose,
-        bool isReleaseBuild,
-        Map<String, Map<String, dynamic>> builderConfigOverrides,
-        List<String> buildDirs}) =>
-    build_impl.build(
-      builders,
-      assumeTty: assumeTty,
+    {bool deleteFilesByDefault,
+    bool failOnSevere,
+    bool assumeTty,
+    String configKey,
+    PackageGraph packageGraph,
+    RunnerAssetReader reader,
+    RunnerAssetWriter writer,
+    Level logLevel,
+    onLog(LogRecord record),
+    Stream terminateEventStream,
+    bool enableLowResourcesMode,
+    Map<String, String> outputMap,
+    bool trackPerformance,
+    bool skipBuildScriptCheck,
+    bool verbose,
+    bool isReleaseBuild,
+    Map<String, Map<String, dynamic>> builderConfigOverrides,
+    List<String> buildDirs}) async {
+  builderConfigOverrides ??= const {};
+  packageGraph ??= new PackageGraph.forThisPackage();
+  var environment = new OverrideableEnvironment(
+      new IOEnvironment(packageGraph, assumeTty, verbose: verbose),
+      reader: reader,
+      writer: writer,
+      onLog: onLog);
+  var options = await BuildOptions.create(environment,
+      configKey: configKey,
       deleteFilesByDefault: deleteFilesByDefault,
       failOnSevere: failOnSevere,
-      configKey: configKey,
       packageGraph: packageGraph,
-      reader: reader,
-      writer: writer,
       logLevel: logLevel,
-      onLog: onLog,
-      terminateEventStream: terminateEventStream,
+      skipBuildScriptCheck: skipBuildScriptCheck,
       enableLowResourcesMode: enableLowResourcesMode,
       outputMap: outputMap,
       trackPerformance: trackPerformance,
-      skipBuildScriptCheck: skipBuildScriptCheck,
       verbose: verbose,
-      builderConfigOverrides: builderConfigOverrides,
-      isReleaseBuild: isReleaseBuild,
-      buildDirs: buildDirs,
-    );
+      buildDirs: buildDirs);
+  var terminator = new Terminator(terminateEventStream);
+  var build = await BuildRunner.create(
```

Comment:
is returning null on failure the right behavior here? What types of reasons might this fail? Would throwing be more appropriate?

Reply:
It now throws.